lib/internal/per_context/safethenable.js

```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+const {
+  PromisePrototypeCatch,
+  PromisePrototypeThen,
+  ReflectApply,
+  SafeWeakMap,
+} = primordials;
+
+const { isPromise } = require('internal/util/types');
+
+const cache = new SafeWeakMap();
+
+// `SafeThenable` should be used when dealing with user provided `Promise`-like
+// instances. It provides two methods `.then` and `.catch`.
+// Wrapping uses of `SafeThenable` in `try`/`catch` may be useful if the
+// accessing `.then` property of the provided object throws.
+class SafeThenable {
+  #is_promise;
+  #makeUnsafeCalls;
+  #target;
+  #cachedThen;
+  #hasAlreadyAccessedThen;
+
+  constructor(thenable, makeUnsafeCalls = false) {
+    this.#target = thenable;
+    this.#is_promise = isPromise(thenable);
+    this.#makeUnsafeCalls = makeUnsafeCalls;
+  }
+
+  get #then() {
+    // Handle Promises/A+ spec, `then` could be a getter
```

Comment:
Why do we care about the Promises/A+ spec in the first place? The only spec that we care about is the JavaScript spec - if we got a native promise this should never happen right?

The "double getteR" thing is an issue (thanks jQuery!) with assimilation mostly.

Reply:
Yes, it is not intended to be used with genuine `Promise` object, it's just for API that have to deal with user-provided _thenable_ objects (E.G.: `util.callbackify`).